governance/api.go

```diff
@@ -146,22 +162,166 @@ func (api *GovernanceKlayAPI) GetRewards(num *rpc.BlockNumber) (*reward.RewardSp
+	TotalBurntTxFee      *big.Int                    `json:"totalBurntTxFee"`
+	TotalProposerRewards *big.Int                    `json:"totalProposerRewards"`
+	TotalStakingRewards  *big.Int                    `json:"totalStakingRewards"`
+	TotalKFFRewards      *big.Int                    `json:"totalKFFRewards"`
+	TotalKCFRewards      *big.Int                    `json:"totalKCFRewards"`
+	Rewards              map[common.Address]*big.Int `json:"rewards"`
+}
+
+// GetRewardsAccumulated returns accumulated rewards data in the range of [start, end).
+// The given block range is divided into many accumulation periods with a given period parameter.
+func (api *GovernanceAPI) GetRewardsAccumulated(start rpc.BlockNumber, end rpc.BlockNumber, period uint64) ([]accumulatedRewards, error) {
+	type request struct {
+		period   uint64
+		blockNum uint64
+	}
 
-// TODO-Klaytn-Mantle: deprecate this
-func (api *GovernanceKlayAPI) ChainConfigAt(num *rpc.BlockNumber) *params.ChainConfig {
-	return getChainConfig(api.governance, num)
-}
+	blockchain := api.governance.BlockChain()
+	govKlayAPI := NewGovernanceKlayAPI(api.governance, blockchain)
 
-func (api *GovernanceKlayAPI) GetChainConfig(num *rpc.BlockNumber) *params.ChainConfig {
-	return getChainConfig(api.governance, num)
+	currentBlock := blockchain.CurrentBlock().NumberU64()
+
+	startBlock := currentBlock
+	if start >= rpc.EarliestBlockNumber {
+		startBlock = uint64(start.Int64())
+	}
+
+	endBlock := currentBlock
+	if end >= rpc.EarliestBlockNumber {
+		endBlock = uint64(end.Int64())
+	}
+
+	if period <= 0 {
+		period = 86400 // one day
+	}
+
+	if startBlock >= endBlock {
+		return nil, errors.New("the end block number should be larger the start block number")
+	}
+
+	if endBlock > currentBlock {
+		return nil, errors.New("the end block number should be equal or less than the current block number")
+	}
+
+	blockCount := endBlock - startBlock
+	if blockCount > 604800 { // 7 days. naive resource protection
+		return nil, errors.New("block range should be equal or less than 604800")
+	}
+
+	// calculate the number of accumulating periods
+	numPeriods := blockCount / period
+	if blockCount%period != 0 {
+		numPeriods++
+	}
+
+	accArray := make([]accumulatedRewards, numPeriods)
+	accLocks := make([]sync.Mutex, numPeriods)
+	rewardArray := make([]*reward.RewardSpec, numPeriods)
+
+	numWorkers := runtime.NumCPU()
+	reqCh := make(chan request, numPeriods) // the size should be equal or larger than the number of request not to block the request loop
+	errCh := make(chan error, 1)
+	wg := sync.WaitGroup{}
+
+	// introduce the worker pattern to prevent resource exhaustion
+	for i := 0; i < numWorkers; i++ {
+		go func() {
+			// the minimum digit of request is period to avoid current access to an accArray item
+			for req := range reqCh {
+				bn := rpc.BlockNumber(req.blockNum)
+				blockReward, err := govKlayAPI.GetRewards(&bn)
+				if err != nil {
+					errCh <- err
```

Comment:
`wg.Done()` should be executed in this branch as well.

FYI, below will panic (all goroutines are asleep).
```go
func main() {
	reqCh := make(chan int, 16)
	numWorkers := 4
	wg := sync.WaitGroup{}

	for i := 0; i < numWorkers; i++ {
		go func(i int) {
			for req := range reqCh {
				if req == 5 {
					return
				}
				fmt.Println(req)
				wg.Done()
			}
			fmt.Println("goroutine", i, "done")
		}(i)
	}

	for i := 0; i < 40; i++ {
		wg.Add(1)
		reqCh <- i
	}
	close(reqCh)

	wg.Wait()
}
```

Reply:
Thanks!